mdagent/tools/base_tools/simulation_tools/setup_and_run.py

```diff
@@ -873,40 +893,144 @@ def _create_system(
 
         # if use_constraint_tolerance:
         #    constraintTolerance = system_params.pop('constraintTolerance')
+        print("About to create system...")
         self.modeller = Modeller(pdb.topology, pdb.positions)
-        if solvate:
-            try:
-                self.modeller.addSolvent(forcefield)
-            except ValueError as e:
-                print("Error adding solvent", type(e).__name__, "–", e)
-                if "No Template for" in str(e):
-                    raise ValueError(str(e))
-            except AttributeError as e:
-                print("Error adding solvent: ", type(e).__name__, "–", e)
-                print("Trying to add solvent with 1 nm padding")
-                if "NoneType" and "value_in_unit" in str(e):
-                    try:
-                        self.modeller.addSolvent(forcefield, padding=1 * nanometers)
-                    except Exception as e:
-                        print("Error adding solvent", type(e).__name__, "–", e)
-                        raise (e)
-            except Exception as e:
-                if "Cannot neutralize the system because the" in str(e):
-                    try:
-                        self.modeller.addSolvent(forcefield, padding=1 * nanometers)
-                    except Exception as e:
-                        print("Error adding solvent", type(e).__name__, "–", e)
+        attempts = 0
+        solvent_list = ["MOH", "EOH", "HOH", "ACN", "URE", "DMS", "DMF", "GOL", "BNZ"]
+        while attempts < 3:
+            print(f"Attempts at creating system: {attempts}/3")
+            if solvate:
+                try:
+                    self.modeller.addSolvent(forcefield)
+                except ValueError as e:
+                    print("Error adding solvent", type(e).__name__, "–", e)
+                    if "No template found for" in str(e):
+                        pattern = r"residue \d+ \((\w+)\)"
+                        # Search for the pattern in the error message
+                        match = re.search(pattern, str(e))
+                        if match:
+                            residue_code = match.group(1)
+                            print(f"Residue code: {residue_code}")
+                        else:
+                            print("No residue code found in the error message.")
+                            raise ValueError(str(e))
+                        if residue_code not in solvent_list:
+                            print(
+                                "Residue code not in solvent list. Adding forcefield \
+                                  not supported."
+                            )
+                            raise ValueError(str(e))
```

Comment:
this if-else block is a bit messy and can be simplified to fewer lines, I think. For an instance, you can use "if" statement only to return all errors: `if no match: .. raise ValueError` and then `if residue_code not in solvent_list:... raise ValueError`. Then keep working on what you want to do. it'd be cleaner and easier to read logic that way.  It's called Early Return: https://medium.com/swlh/return-early-pattern-3d18a41bba8 -- and I love that funny quote in the intro fig in that article lol

Reply:
From what you wrote here, it seems that small molecules are always treated as solvent. does that work for standalone small molecules as well? Is that what we want?

Reply:
is not that small molecules are always treated as solvents. Is that based on our conversation, we decided to just worked on solvents for now, and avoid protein ligands interactions. @SamCox822 right?

Reply:
protein-ligand interaction is a separate thing, for sure. I was referring to molecules only vs. molecules + proteins. As we discussed, you confirmed it can do both scenarios. I still think this block full of nested conditions (line 905-942) could be shorter and more readable lol